### ISSUE TITLE
pyright hangs while using 100% CPU

### ISSUE THREAD

**Issue by bkoropoff** [microsoft/pyright#4475]
**Describe the bug**
pyright hangs and uses 100% CPU on particular python code.

**To Reproduce**
Running `pyright <file.py>` on the code below hangs and pegs a CPU at 100%.

**Expected behavior**
It doesn't do that.

**Screenshots or Code**
This is the most reduced test case I could produce:
```python
from typing import Iterable, TypeVar, Union

T = TypeVar("T")
Tree = list[Union["Tree[T]", T]]

def _flatten(tree: Union[Tree[T], T]) -> Iterable[T]:
    if not isinstance(tree, list):
        yield tree
        return
    for v in tree:
        yield from _flatten(v)

def flatten(tree: Tree[T]) -> Iterable[T]:
    return _flatten(tree)
```
**VS Code extension or command-line**
Command line, latest version from pip (1.1.289), Linux.

**Additional context**
Originally found using pyright as LSP from neovim.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by bkoropoff** [comment]
The latest version from git still hangs if I modify the test case to call the function:

```python
from typing import Iterable, TypeVar, Union

T = TypeVar("T")
Tree = list[Union["Tree[T]", T]]

def _flatten(tree: Union[Tree[T], T]) -> Iterable[T]:
    if not isinstance(tree, list):
        yield tree
        return
    for v in tree:
        yield from _flatten(v)

def flatten(tree: Tree[T]) -> Iterable[T]:
    return _flatten(tree)

flatten([1,[2,3]])
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
